File: DataStructureJava/array/heap/HeapSort.java
public class HeapSort {
    private HeapSort(){}

    public static <E extends Comparable<E>> void sort(E[] data){ // data 是一个无序数组
        // 把data heapify 成一个最大堆
        if (data.length <= 1) return;
        for (int i = (data.length-2) / 2; i >= 0; i --)
            siftDown(data,i,data.length);

        for (int i = data.length - 1; i >= 0; i--){
            swap(data,0,i);
            siftDown(data,0,i);
        }
    }
    // 对[0,n)所形成的最大堆中，索引 k 的元素，执行siftdown
    private static <E extends Comparable<E>> void siftDown(E[] data, int k , int n){
        while (2 * k + 1 < n){
            int j = 2 * k + 1;
            // 如果右孩子的值大于左孩子的值，
            if (j + 1 < n && data [j + 1].compareTo(data[j]) > 0)
                j ++; // j = rightChild(k)
            // 此时 j 是左右孩子中的最大值
            if (data[k].compareTo(data[j]) >= 0)
                break;
            swap(data,k,j);
            k = j;
        }
    }
    public static <E> void  swap(E[] data,int i, int j) {

        E t = data[i];
        data[i] = data[j];
        data[j] = t;
    }
}

File: data_structure_python/heap/heap_sort.py
# 用最大堆实现排序

class HeapSort:
    def __init__(self, array):
        self.arr = array

    def sort(self):
        if len(self.arr) <= 1:
            return
        for i in range((len(self.arr) - 2) // 2, -1, -1):  # heapify的过程
            self.sift_down(i, len(self.arr))

        for i in range(len(self.arr) - 1, -1, -1):
            self.arr[0], self.arr[i] = self.arr[i], self.arr[0]
            self.sift_down(0, i)

    # 对[0,n)所形成的最大堆中，索引k的元素，执行sift_down
    def sift_down(self, k, n):
        while 2 * k + 1 < n:  # 左孩子的索引小于size
            i = 2 * k + 1  # i = 左孩子索引
            if i + 1 < n and self.arr[i + 1] > self.arr[i]:
                i += 1  # i 是左右孩子的大的那个
            if self.arr[k] >= self.arr[i]:
                break
            self.arr[k], self.arr[i] = self.arr[i], self.arr[k]

            k = i  # k指向被换后的位置,继续判断是否需要sift_down

    def __str__(self):
        return str(self.arr)


if __name__ == '__main__':
    ll = [2, 3, 1, 0, 5, 4]
    hq = HeapSort(ll)
    hq.sort()
    print(hq)
